aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -1042,4 +1042,21 @@ Tensor dist(const Tensor &self, const Tensor& other, Scalar p){
   return at::norm(self - other, p);
 }
 
+bool cpu_equal(const Tensor& self, const Tensor& other) {
+  if (!at::namedinference::are_names_equal(
+        self.unsafeGetTensorImpl(), other.unsafeGetTensorImpl())) {
+    return false;
+  }
+  at::NoNamesGuard guard;
+  TORCH_CHECK(self.device() == other.device(), "Cannot compare two tensors on "
+              "different devices. Got: ", self.device(), " and ", other.device());
+  TORCH_CHECK(self.dtype() == other.dtype(),
+              "Expected object of scalar type ", self.dtype(), " but got scalar type ",
+              other.dtype(), " for argument 'other'");
+  if (!self.is_same_size(other)) {
+    return false;
+  }
+  return at::native::eq(self, other).all().item().to<bool>();
```

Comment:
@xcnick 
Based on your benchmark, I found one problem when calling by calling at::native::eq directly.
for contiguous non equal test, the performance of the at::native::eq is very bad, the reason is that, at::native::eq will compare every element between the two tensors, even an unequal case is already found, it won't stop. This will also slow down the non-contiguous non equal case, but not as bad as contiguous case.

Let's do something else here:
Instead of calling eq directly, let's implement our own loop function. You can put the following code here and remove the 'return at::native::eq' line.
```
  bool result = true;
  auto iter = TensorIteratorConfig()
    .add_input(self)
    .add_input(other)
    .allow_cpu_scalars(true)
    .promote_inputs_to_common_dtype(true)
    .build();

  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBool, kBFloat16, kHalf, iter.input_dtype(), "equal_cpu", [&]{
    iter.for_each([&](char** data, const int64_t *strides, int64_t dim_size){
      if (!result) {
          return;
      }
      scalar_t* self_data = (scalar_t*)data[0];
      scalar_t* other_data = (scalar_t*)data[1];
      for (int64_t i = 0; i < dim_size; ++i) {
        if (self_data[i] != other_data[i]) {
            result = false;
            break;
        }
      }
    });
  });
  return result;
```

Let me explain a little bit. Since we just need to return a bool, so there is no need to even have an output tensor. So we initialize a tensor iterator with two input only, and keep track a global boolean 'result'

The tensor iterator has a for_each function which takes a function reference.
This reference has signatures like this: loop(char** data, const int64_t* strides, int size);
Tensor iterator internally, chunk the whole tensor into many 2D planes, and process them row by row, this loop function is used to iterator each row. 'data' array contains the starting point of current row for each input tensor, 'strides' array contains stride of current row (dimension) for each input tensor, the tensor orders for both 'data' and 'strides' are the same as what you initialize the tensor iterator. And the 'size' is the row length.

This code dispatch the iter.for_each() to all the supported data type, and if there is any non-equal element found between two tensors, it stopped immediately. For contiguous tensor, this means quit the whole tensor iteration since we totally have one continuous data array in this case. If tensor is not contiguous, there is still overhead, since in this case, the program will jump to next row and immediately break, and jump to next row, break, until looped entire tensors. But it already saved many time compare calling at::native::eq directly, and plus there is no tensor write. 

Please make this change in and update the benchmark and rebase, I think we are good to go then.